### PR TITLE
Override node name with a name containing underscore

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Fixed https://github.com/flyteorg/flyte/issues/3608
Convert the node name into a DNS-compliant.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3608

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by pingsutw** [comment]
cc @nan2018

File: tests/flytekit/unit/core/test_node_creation.py [hunk 1]
```diff
@@ -427,6 +427,7 @@ def my_wf(a: str) -> str:
     wf_spec = get_serializable(OrderedDict(), serialization_settings, my_wf)
     assert len(wf_spec.template.nodes) == 1
     assert wf_spec.template.nodes[0].metadata.name == "foo"
+    assert wf_spec.template.nodes[0].id == "t-1"
```
Comment (Nan2018) [inline tests/flytekit/unit/core/test_node_creation.py:430]: why dash instead of underscore?

Comment (pingsutw) [reply tests/flytekit/unit/core/test_node_creation.py:430]: underscore can't be used in pod name. https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-s-dns-policy

File: flytekit/core/node.py [hunk 2]
```diff
@@ -84,7 +84,7 @@ def metadata(self) -> _workflow_model.NodeMetadata:
 
     def with_overrides(self, *args, **kwargs):
         if "node_name" in kwargs:
-            self._id = kwargs["node_name"]
```
Comment (eapolinario) [inline flytekit/core/node.py:87]: can you leave a comment here pointing to https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names as the spec for pod names? 

Also, worth noting that node names usually map to pod names, but not always. We don't have worry about this here though (i.e. it's fine to have `_dnsify` here).

**Review by eapolinario — APPROVED** [review]

**Pull request merged**